url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/LegacyOCPMessage.java

```diff
@@ -0,0 +1,436 @@
+        return messageTypes.get(commandCode);
+    }
+
+    /**
+     * Lookup the OCPMessage implementation for a specific command code.
+     *
+     * @param commandCode
+     *            the command code from the OCP message
+     * @return the Class that handles the specified command code
+     */
+    public static Class<? extends LegacyOCPMessage> getOCPClass(
+            final short commandCode) {
+        LegacyOCPMessageTypes type = getOCPType(commandCode);
+
+        if (type != null) {
+            return type.getImplementation();
+        } else {
+            return null;
+        }
+    }
+
+    //CSOFF: JavadocMethod
+    /**
+     * Decode a binary OCP message into an instance of the implementing class.
+     *
+     * @param buffer
+     *            an OCP message
+     * @return a subclass of OCPMessage that handles the binary message
+     * @throws OCPException
+     *             if an error occurs while decoding the message
+     * @throws UnknownMessageTypeException
+     *             if the command type is not recognised
+     * @throws LinkMessageException
+     *             if the command code is that of an unimplemented link command
+     * @throws CallMessageException
+     *             if the command code is that of an unimplemented call command
+     */
+    public static LegacyOCPMessage decodeBuffer(final ByteBuffer buffer)
+            throws OCPException {
+        if (buffer.limit() < OCP_MIN_LENGTH) {
+            // Too short to possibly be a valid message
+            throw new OCPException();
+        }
+
+        Class<? extends LegacyOCPMessage> implementation;
+        Constructor<? extends LegacyOCPMessage> constructor;
+
+        short commandCode = buffer.getShort(OCP_CMD_CODE_OFFSET);
+        implementation = getOCPClass(commandCode);
+        if (implementation == null) {
+            // Unimplemented OCP message. Try and work out what sort it was.
+            switch (commandCode & OCP_COMMAND_TYPE_MASK) {
+            case OCP_COMMAND_TYPE_LINK:
+                throw new LinkMessageException(
+                        commandCode,
+                        LinkCommandUnsupported.REASON_COMMAND_CODE_UNSUPPORTED);
+            case OCP_COMMAND_TYPE_CALL:
+                // Extract the task IDs from the original message
+                throw new CallMessageException(
+                        buffer.getInt(LegacyOCPMessage.OCP_DEST_TID_OFFSET),
+                        buffer.getInt(LegacyOCPMessage.OCP_ORIG_TID_OFFSET),
+                        commandCode,
+                        CallCommandUnsupported.REASON_COMMAND_CODE_UNSUPPORTED);
+            default:
+                throw new UnknownMessageTypeException();
+            }
+        }
+
+        try {
+            constructor = implementation.getConstructor(ByteBuffer.class);
+            return constructor.newInstance(buffer);
+        } catch (InvocationTargetException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)